File: backend/app/domain/entities/chat_turn_summary.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime

from backend.app.domain.entities.agent_configuration import AgentId, AgentType
from backend.app.domain.entities.chat_message import (
    ChatMessage,
    ChatMessageAuthorType,
    ChatMessageRole,
    ChatMessageStatus,
)

from backend.app.domain.entities.job import utc_now


@dataclass(frozen=True, slots=True)
class ChatTurnSummarySourceMessage:
    message_id: str
    role: ChatMessageRole
    author_type: ChatMessageAuthorType
    agent_id: AgentId
    agent_type: AgentType
    agent_label: str | None
    content: str | None
    status: ChatMessageStatus
    created_at: datetime

    @classmethod
    def from_message(cls, message: ChatMessage) -> "ChatTurnSummarySourceMessage":
        return cls(
            message_id=message.id,
            role=message.role,
            author_type=message.author_type,
            agent_id=message.agent_id,
            agent_type=message.agent_type,
            agent_label=message.agent_label,
            content=message.content,
            status=message.status,
            created_at=message.created_at,
        )


@dataclass(slots=True)
class ChatTurnSummary:
    id: str
    session_id: str
    content: str
    source_message_ids: tuple[str, ...] = ()
    source_messages: tuple[ChatTurnSummarySourceMessage, ...] = ()
    created_at: datetime = field(default_factory=utc_now)
    updated_at: datetime = field(default_factory=utc_now)

File: frontend/mobile_app/lib/src/models/chat_turn_summary.dart
import 'agent_configuration.dart';
import 'chat_message.dart';

class ChatTurnSummarySourceMessage {
  const ChatTurnSummarySourceMessage({
    required this.messageId,
    required this.isUser,
    required this.authorType,
    required this.agentId,
    required this.agentType,
    this.agentLabel,
    this.content,
    required this.status,
    required this.createdAt,
  });

  final String messageId;
  final bool isUser;
  final ChatMessageAuthorType authorType;
  final AgentId agentId;
  final AgentType agentType;
  final String? agentLabel;
  final String? content;
  final ChatMessageStatus status;
  final DateTime createdAt;

  factory ChatTurnSummarySourceMessage.fromJson(Map<String, dynamic> json) {
    return ChatTurnSummarySourceMessage(
      messageId: json['message_id'] as String,
      isUser: (json['role'] as String?) == 'user',
      authorType: _turnSummaryAuthorTypeFromJson(
        json['author_type'] as String?,
        role: json['role'] as String? ?? 'assistant',
      ),
      agentId: agentIdFromJson(json['agent_id'] as String? ?? 'generator'),
      agentType: agentTypeFromJson(json['agent_type'] as String? ?? 'generator'),
      agentLabel: json['agent_label'] as String?,
      content: json['content'] as String?,
      status:
          _turnSummaryStatusFromJson(json['status'] as String? ?? 'completed'),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class ChatTurnSummary {
  const ChatTurnSummary({
    required this.id,
    required this.content,
    this.sourceMessageIds = const <String>[],
    this.sourceMessages = const <ChatTurnSummarySourceMessage>[],
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String content;
  final List<String> sourceMessageIds;
  final List<ChatTurnSummarySourceMessage> sourceMessages;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory ChatTurnSummary.fromJson(Map<String, dynamic> json) {
    final rawSourceMessageIds =
        json['source_message_ids'] as List<dynamic>? ?? <dynamic>[];
    final rawSourceMessages =
        json['source_messages'] as List<dynamic>? ?? <dynamic>[];
    return ChatTurnSummary(
      id: json['id'] as String,
      content: json['content'] as String,
      sourceMessageIds: rawSourceMessageIds
          .whereType<String>()
          .toList(growable: false),
      sourceMessages: rawSourceMessages
          .whereType<Map<String, dynamic>>()
          .map(ChatTurnSummarySourceMessage.fromJson)
          .toList(growable: false),
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

ChatMessageAuthorType _turnSummaryAuthorTypeFromJson(
  String? authorType, {
  required String role,
}) {
  switch (authorType) {
    case 'assistant':
      return ChatMessageAuthorType.assistant;
    case 'reviewer_codex':
      return ChatMessageAuthorType.reviewerCodex;
    case 'human':
      return ChatMessageAuthorType.human;
    default:
      return role == 'assistant'
          ? ChatMessageAuthorType.assistant
          : ChatMessageAuthorType.human;
  }
}

ChatMessageStatus _turnSummaryStatusFromJson(String status) {
  switch (status) {
    case 'reserved':
      return ChatMessageStatus.reserved;
    case 'submission_pending':
      return ChatMessageStatus.submissionPending;
    case 'submission_unknown':
      return ChatMessageStatus.submissionUnknown;
    case 'pending':
      return ChatMessageStatus.pending;
    case 'failed':
      return ChatMessageStatus.failed;
    case 'cancelled':
      return ChatMessageStatus.cancelled;
    case 'sending':
      return ChatMessageStatus.sending;
    default:
      return ChatMessageStatus.completed;
  }
}
